docs/enhancements/support-node-labeling.md

```diff
@@ -0,0 +1,159 @@
+for the role.  This MCP has the format `rendered-<role>-*` (* stands for a hash string).  So there is MCP for master 
+role `rendered-master-*` and for worker `rendered-worker-*`.
+
+When adding custom role, it can be associated with custom MCP which references specific configuration
+for that role.  Setting appropriate label a node, causes the node to reboot and apply the 
+MCP configuration (machine config) on that node.
+
+Setting node labels should support this flow.
+
+### Goals
+- For ZTP flow: add capability to add and update annotations on BMH which will be transferred 
+to agent CR.
+- For boot it yourself flow: add capability to add and update labels to agent CR after agent
+  has been created.
+- Transfer the labels from the agent CR to the corresponding node on the spoke cluster.
+
+### Non goals
+- Single way to set node labels before agent creation both for ZTP flow, and boot it yourself flow.
+
+## Proposal
+
+### ZTP flow
+Labels can be applied on every CR.  So there is no known limitation to add a label on agent.
+
+For the ZTP flow, there should a pattern for transferring an annotation from the BMH to a label  
+in the corresponding agent.  The pattern is actually the prefix of the annotation.
+It should be **node-label**.  The suffix of the annotation will be transferred 
+to labels on the corresponding agent CR.
+
+For example assuming we want to apply a label **node-role.kubernetes.io/infra** on an
+a node.  Then the following annotation should be added to the BMH: **node-label.node-role.kubernetes.io/infra**.
+
+### Node labels on agent CR
+Additional field **nodeLabels** will be added to the spec 
+section of the agent CR.  This field will be a string list.
+
+For the ZTP flow, the annotations having the prefix **node-label** in BMH will be transferred
+to the dedicated field **nodeLabels** without the **node-label** prefix.
+
+If boot it yourself flow is in use, nodeLabels should be set directly on the agent CR.
+
+### Moving labels to nodes
+
+#### Day 1 cluster
+When installing day 1 cluster, the implementation should make use of the **nodeLabels** field
+which is part of host DB record. Once the node becomes ready, the label is applied on the node.
+This functionality already exists in the assisted controller which runs as a pod on the 
+spoke cluster during installation.
+
+#### Day 2 nodes
+The flow can be supported only when working with Kubernetes API, since direct access to
+the cluster is required.  In this case, setting node labels should be done by a 
+Kuberenetes controller (probably agent controller).
+
+In addition, for day 2 nodes we need to support setting custom manifests.  This functionality is
+already supported for day 1 clusters by the assisted controller.  In order to allow adding new
+MCP to existing day 2 clusters, this functionality needs to be supported by a Kubernetes controller - 
+probably the cluster-deployment controller.
```

Comment:
Adding MCPs to existing clusters feels out of scope for Assisted.

Reply:
OK.